### ISSUE TITLE
[Feature Request] Mouse and controller mappings for lightgun type game controllers for light gun shooter games?

### ISSUE THREAD

**Issue by Hedda** [kodi-game/controller-topology-project#171]
@garbear Could the Controller Topology Project also map mouse (X pointer) + buttons from modern lightgun controllers map to buttons on retro "light gun" controllers for emulation?

Classic Light gun accessories are the [NES Zapper](https://en.wikipedia.org/wiki/NES_Zapper), the Light Phaser for the [Sega Master System](https://en.wikipedia.org/wiki/Sega_Master_System), and the [Atari](https://en.wikipedia.org/wiki/Atari_Corporation) [XG-1](https://en.wikipedia.org/wiki/XG-1)

https://en.wikipedia.org/wiki/Light_gun

![image](https://user-images.githubusercontent.com/6320001/194536226-250359bc-f44d-48b1-ac53-e1cf61a24b0d.png)      ![image](https://user-images.githubusercontent.com/6320001/194536257-e7ead0ef-fbb1-4afd-9a3e-0f0ae03feced.png)   ![image](https://user-images.githubusercontent.com/6320001/194536273-a1d96a9c-ca99-4749-aa33-8c395f07f0f9.png)

With lightguns, it would be possible to play some or all classic Lightgun Shooter games from both game consoles and the arcade.

https://en.wikipedia.org/wiki/Light_gun_shooter

https://en.wikipedia.org/wiki/List_of_gun_games

Lightgun Shooter popular favourite examples such as:

- Duck Hunt
- Wild Gunman
- Time Crisis 1/2/3
- Virtua Cop 1/2/3
- House Of The Dead 1/2/3/4/Overkill
- Point Blank 1/2/3
- Terminator 2 Arcade
- Star Wars Trilogy
- Jurassic Park: Lost World
- Mad Dog McCree
- Battle Clash
* And many many more lightgun games.

As you probably know, the Nintendo Wiimote + Wii Sensor Bar combination is a relatively popular hardware choice to achieve an inexpensive wireless lightgun for emulators on PC computers, as a bonus there are loads of different Zapper Gun holder accessories for the Nintendo Wiimote, though it should be noted that Wiimote is not as fast/good as a modern light gun when only combined with a standard Wii Sensor Bar.

https://www.instructables.com/Accurate-Wiimote-Light-Gun-on-Raspberry-PI/

Today there are also multiple other light guns options that were made for, and unlike classic lightguns which only works on CRT monitors these newer lightguns work with modern LCD/OLED displays by using some different technologies:

Check out for example; Ultimarc Aimtrak Light Gun", "Sinden Lightgun", and "PDP Mars Lightgun" ("PDPMars) lightguns:

- https://www.ultimarc.com/aimtrak.html
  * Ultimarc Aimtrak Light Gun independent fan review and collected information + video demos -> https://lightgungamer.com/aimtrak-light-gun/

- https://sindenlightgun.com
  * Sinden Lightgun independent fan review and collected information + video demos -> https://lightgungamer.com/sinden-lightgun/
  * Primarily meant for a premium arcade experience and has loads of accessories -> https://www.sindenshop.com

- https://mars.pdp.com
  * PDP Mars Lightgun ("PDPMars) independent fan review and collected information + video demos -> https://lightgungamer.com/pdp-mars-lightgun/
  * This can today be bought bundled with Xbox One or PlayStation 4 / PlayStation 5 Games = https://www.amazon.com/s?k=PDP+MARS&ref=nb_sb_noss

![image](https://user-images.githubusercontent.com/6320001/194536524-b82b89fc-4147-43e7-961d-a3f69cb64510.png)

![image](https://user-images.githubusercontent.com/6320001/194536543-98951876-7442-4b31-9d7e-d1a8409d68db.png)

![image](https://user-images.githubusercontent.com/6320001/194536565-5a140beb-d39f-4c2b-99a4-631e7fec0e50.png)

**Comment by Hedda** [comment]
GUN4IR by JB based on reviews sounds like it could be an awesome affordable (and highly customizable) DIY lightgun system for arcade experience:

https://www.gun4ir.com

GUN4IR hardware is closed source but sofytware is open, though availability for buying the DIY version seems scarce since only manufactured in samller batches 

http://forum.arcadecontrols.com/index.php/topic,161189.0.html

He also partners with some othes to sell pre-modded GUN4IR lightguns, like example

https://rpegelectronics.com/products/gun4ir-light-gun-ready-to-play-namco-gcon45-black?variant=40080266461373

PS: GUN4IR is made by Jean-Baptiste Bongrand (a.k.a. JayBee) who also makes some other open-source stuff so maybe that will change in the future(?)

https://github.com/JayBee-git

**Comment by garbear** [comment]
I love supporting independent projects like GUN4IR! X-Arcade sent me a Tankstick for free and I feel back it took 2 years to finish the driver. GUN4IR is something I would looooove to buy and get working in Kodi, but I just don't have time these days. Life changes though, and someday I may be able to support projects like this.

**Comment by Hedda** [comment]
> I love supporting independent projects like GUN4IR!

@JayBee-git maybe you would consider looking into coding a GUN4IR compatible implementation for Kodi’s native game framework support yourself?

I noticed that you submitted C++ code for adding native Gun4IR lightgun input support for MiSTer so you clearly have the needed coding skills -> https://github.com/MiSTer-devel/Main_MiSTer/pull/633

For reference if you have not heard about Kodi (formely XBMC) and its native game support with ”RetroPlayer” (using Libreteo), read:

https://kodi.wiki/view/Game_development

https://kodi.wiki/view/Game_controllers

https://forum.kodi.tv/showthread.php?tid=292270

https://forum.kodi.tv/showthread.php?tid=195552

https://forum.kodi.tv/showthread.php?tid=257764

https://forum.kodi.tv/showthread.php?tid=230276

https://forum.kodi.tv/showthread.php?tid=252465

https://forum.kodi.tv/showthread.php?tid=164725

https://forum.kodi.tv/showthread.php?tid=340684

**Comment by garbear** [comment]
And @JayBee-git if you're interested in writing a Kodi driver you have my full support and code review!

**Comment by JayBee-git** [comment]
Hey, I'd gladly help if I can.
But for the MiSTer it was very easy, as they already had a driver and everything in place to support lightguns. It was only a few lines of code to add.
Lightgun inputs are nothing more than generic mouse, keyboard and joystick inputs, so as long as you have a generic driver that supports any of these, it should be quite easy to add. The only thing special compared to normal inputs is that they are using absolute axis for the mouse input instead of relative.
I'm not sure what was implemented already in Kodi, but I sadly don't have any knowledge on how libretro drivers work and how to code them.
If you need any info on how the lightgun inputs work tho, feel free to ask me.

**Comment by garbear** [comment]
Oh no! Mouse and keyboard controlling joystick games is a feature we're missing. I implemented this once but had to remove it to simplify the system.

I plan to bring the feature back in a "Player Manager". You can see the saga here: https://github.com/garbear/xbmc/issues/87

Most of the code is in place, we just don't have the final step that converts the input because we don't have a way to map it in the UI.

However, I was able to get the tankstick to work by reading the key presses from Linux and then sending joystick buttons. That work is here: https://github.com/kodi-game/peripheral.xarcade/

You could do the same for keyboard and mouse input, get it directly from Linux in a small peripheral add-on. It would only work on Linux but that covers the whole LibreELEC community which is huge.

**Comment by JayBee-git** [comment]
The gun also work in a joystick input mode, the one that is used by MiSTer. You can map games lightgun axis to the joystick input axis directly.

**Comment by JayBee-git** [comment]
Sure, which version do you want me to try out?

**Comment by garbear** [comment]
Version 20.0. I actually merged a little bit of the player manager into v20. Baby steps.

**Comment by JayBee-git** [comment]
I meant which OS, machine, and where to get it?
I know Kodi but I never used Kodi game 😄

**Comment by garbear** [comment]
Its built into Kodi so games should work everywhere that has emulators in the addon browser

**Comment by Hedda** [comment]
@JayBee-git  FYI, Kodi RetroPlayer (i.e. Kodi Game,) runs on most operating systems, processor Instruction set architectures and computer platforms available today, however, I believe that for development purposes it is best to use an Intel/AMD x86-64 based computer with either Linux or Windows operating system (depending on your personal preference as a developer).

- https://en.wikipedia.org/wiki/Kodi_(software)

  - https://en.wikipedia.org/wiki/Kodi_(software)#Games_Manager_and_game_console_emulators

Links and information on where to download and how to install Kodi 20.x (codename "Nexus") releases can be found here:

-https://kodi.tv/download/

Alternatively, you could get garbear's very latest custom test releases if just want to test out all the existing features:

-https://forum.kodi.tv/showthread.php?tid=173361 -> https://github.com/garbear/xbmc/releases

-Here is a guide on how to get started with Kodi RetroPlayer:

-https://forum.kodi.tv/showthread.php?tid=340684

Console emulators are available as "game clients" add-ons, and joysticks/gamepad mappings are available as "Controller profiles":

- https://kodi.tv/addons/nexus/ 

  - https://kodi.tv/addons/nexus/category/game-clients
  
  - https://kodi.tv/addons/nexus/category/controller-profiles

Also check out the Kodi wiki section on how Games and game development work in Kodi here:

-https://kodi.wiki/view/Games

PS: If you used Kodi from an end-users perspective then you will notice that Kodi's GUI is missing a games-library (and database).

**Comment by JayBee-git** [comment]
@Hedda Yes, actually I do have kodi already installed on most of my machines :)
My bad, I was thinking Kodi Game was a fork not an addon.
I will test it out when I have the time, and tell you how it went.

About the donation of a gun, it's important to note that I'm not a company or anything (yet), just working on this diy oriented system on my free time. I only do very occasional pre modded gun productions myself (only did one so far), and all the money I get from the small donations is used in gun4ir related expenses.
So, long story short, I sadly don't have the funds to send a free 200$ gifts to devs.

**Comment by JayBee-git** [comment]
That being said I'd happy to help if any of you want to build a gun4ir, and I'll help any way I can.
It's fun to bring back gaming to Kodi, reminds me of the good old xbmc days on the og xbox.

**Comment by Hedda** [comment]
> My bad, I was thinking Kodi Game was a fork not an addon.

Not quite, it is not an add-on and it is a little more complicated than that ;)

Just to try to summerize; "Game API" (with "Kodi RetroPlayer" + "Game Clients" + "Libretro Wrapper") and "Hardware API" (with "Input Devices" + "controller profiles") is now natively integrated subsystems inside the upstream mainline Kodi codebase at https://github.com/xbmc/xbmc/ 

"Kodi Game" is more an internal/ informal umbrella name for all Kodi game-related extensions that are able to extend the native features and functions, and that part does include official add-ons such "game clients" and "controller profiles". So  "Kodi Game" is not add-on, but it several add-ons are part of  "Kodi Game".

https://github.com/orgs/kodi-game/repositories

garbear does a better job at explaining in this forums thread where he also drawn a diagram how it all fits togther inside Kodi:

https://forum.kodi.tv/showthread.php?tid=195552

![image](https://user-images.githubusercontent.com/6320001/216436873-12bd8dc8-8118-4c56-b7b5-1d16ef83e2e3.png)

_"this is the current layout of the retroplayer project. it is centered around two APIs, the Game API (connects retroplayer to game clients) and the Hardware API (connects game clients to input). the Hardware API moves all of the joystick code out of XBMC and into binary add-ons."_

There is also a summarized overview in the official Kodi wiki here:

https://kodi.wiki/view/Game_development

PS: By now natively supporting Libretro game emulator cores via a libretro wrapper (libretro API), Kodi is part of a larger libretro software ecosystem that enables retro gaming which several different independent software projects make use of by Libretro game emulator cores:

https://www.libretro.com/index.php/api/

https://forum.kodi.tv/showthread.php?tid=292270

**Comment by garbear** [comment]
@JayBee-git I'll make you a deal. If you can get the lightgun working enough in Kodi I'll buy the best model you got :slightly_smiling_face:

**Comment by JayBee-git** [comment]
> @JayBee-git I'll make you a deal. If you can get the lightgun working enough in Kodi I'll buy the best model you got 🙂

You do realize I have nothing to sell myself, right?
As I said my system is first and foremost a DIY system, if you want to support it, you can easily build it 😉

And also, even if I wanted, making a libretro driver isn't something I can do like this out of the blue, it's not my field at all, unless lightgun support has already been fully implemented by the devs.

Plus there are many questions that need to be asked before even considering adding it. How are input handled in the hardware API exactly (direct raw hid? through the OS generic joypad/mouse hid input? Through kodi own input layer?)? What are the emulated input types (not the hardware side) supported inside kodi libretro at the moment? How is the measured latency added by the API? Does any of the cores ported have support for lightgun?

I'll test it out when I have time, and see what I can do, but I can't make any promises.

**Comment by garbear** [comment]
> You do realize I have nothing to sell myself, right?

I saw the pre-modded GUN4IR sets in the shop. Are those the model you're trying to get working in Kodi, or is it your DIY kit?

I see what you're saying though, it's still early days.

> How are input handled in the hardware API exactly (direct raw hid? through the OS generic joypad/mouse hid input? Through kodi own input layer?)?

Interfacing with input APIs is all what the peripheral.joystick add-on is about. The history:

Early in Kodi's history, massive amounts of duplication and dependency spew that had organically emerged were slowing down the project, so I moved all input APIs into the add-on. It wasn't ideal, because add-ons are supposed to be optional now we have a "mandatory" add-on. However, it encapsulated all input APIs and got them out of core, so I considered it a win.

You can see all the input APIs in the [build file](https://github.com/xbmc/peripheral.joystick/blob/Nexus/CMakeLists.txt).

Currently this list includes:

* SDL2 (but only on the SteamLink, an abandoned port)
* Cocoa
* Linux Joystick API
* DirectInput
* XInput
* udev

Because the input APIs of Android, iOS and tvOS have the much heavier entire platform as a dependency, they've been implemented in core instead of the add-on.

The next question is, what are all the USB/otherwise protocols that the hardware supports?

**Comment by garbear** [comment]
I added Lightgun Support as a GSoC 2023 project, in case a student out there wants to work on getting Lightguns working over the summer: https://kodi.wiki/view/Google_Summer_of_Code/2023#Lightgun_support

**Comment by JayBee-git** [comment]
> I saw the pre-modded GUN4IR sets in the shop. Are those the model you're trying to get working in Kodi, or is it your DIY kit?
To quote myself from my previous message:
> About the donation of a gun, it's important to note that I'm not a company or anything (yet), just working on this diy oriented system on my free time. I only do very occasional pre modded gun productions myself (only did one so far), and all the money I get from the small donations is used in gun4ir related expenses.

To sum up, I rarely have anything to sell, and when I do, it's in very limited stock, and the price I have is just enough to cut my loss. I only do sales for my community, not to make money, as strange as it might sound. So while I appreciate the offering to buy my gun, these are specifically made for users that really want to use my system but don't have the skills to build it themselves. Selling you a premodded lightgun that I am not sure you'll even use, would be taking one away from someone else who really needs it ;)
All Gun4IR lightguns are pretty much the same as the DIY, there is only one Gun4IR system so far.

> I see what you're saying though, it's still early days.
Not really, I have been doing that for 3 years now. And the gun is pretty much plug and play on most systems already (windows, linux, retropie, mister...). But again, on the hardware side the Gun4IR is still a DIY system only, not a commercial product yet.

> Currently this list includes:
> 
> * SDL2 (but only on the SteamLink, an abandoned port)
> * Cocoa
> * Linux Joystick API
> * DirectInput
> * XInput
> * udev
> 
> Because the input APIs of Android, iOS and tvOS have the much heavier entire platform as a dependency, they've been implemented in core instead of the add-on.
Cool, then you already have everything needed to make the gun work 😄 

> The next question is, what are all the USB/otherwise protocols that the hardware supports?
I will quote myself again;
> Lightgun inputs are nothing more than generic mouse, keyboard and joystick inputs
and
> The gun also work in a joystick input mode, the one that is used by MiSTer. You can map games lightgun axis to the joystick input axis directly.

There is nothing special about lightgun inputs, if your system cores and emulated input system already support lightguns with analog stick axis or absolute mouse axis, it will just work out of the box. You can test it out by yourself without lightguns, by mapping any dinput/xinput analog stick to the lightgun axis. If that works, my lightgun will also work.

**Comment by Hedda** [comment]
> I added Lightgun Support as a GSoC 2023 project, in case a student out there wants to work on getting Lightguns working over the summer: https://kodi.wiki/view/Google_Summer_of_Code/2023#Lightgun_support

@garbear not that I have the skills to program that myself but for reference wondering if the code for different lightguns should preferably go into the main peripheral.joystick driver project at https://github.com/xbmc/peripheral.joystick or into a separate driver like the X-Arcade Tankstick driver project? https://github.com/kodi-game/peripheral.xarcade or somewhere else?

**Comment by Hedda** [comment]
> for the MiSTer it was very easy, as they already had a driver and everything in place to support lightguns. It was only a few lines of code to add.

@garbear  Oh, and I guess that the input code from MiSTer would be of no help? https://github.com/MiSTer-devel/Main_MiSTer

**Comment by Hedda** [comment]
FYI, here is a "Lightguns" YouTube playlist by Retro Ralph who reviews Lightgun hardware (and games) for retro game emulation:

https://www.youtube.com/watch?v=QPmmFR3bBMw&list=PLH48XsrRH8eFxozpHSYaI4iBZsIbFuKLq&ab_channel=RetroRalph

**Comment by Hedda** [comment]
FYI, just watched this video from Retro Ralph about a new low-cost lightgun product called "**_Retro Shooter_**" (or "**_RetroShooter"_**):

https://www.youtube.com/watch?v=J_Gp5HtfsA0&ab_channel=RetroRalph

 "**_Retro Shooter"_** (model "RS-5" or "RS5") is a new budget lightgun hardware plug-and-play solution.

https://retroshooter.com

Interestingly by the sounds of their back-story they have similar thoughts on the topic of lightgun games as discussed above.

https://retroshooter.com/back-story/

Produced in partnership/collaberation with the people behind Pandora Box, it is brand new and currently only ships bundled with a new closed-source Pandora Box based emulation console, but it does also come with a USB breakout-box ("Fire Base" for the Pandora Box 10), which unfortunately is not yet sold separately, so if and when they release a guns-only package with only lightguns, pedals and sensors + a USB breakout-box for the Pandora Box 10 instead of the full bundle with a console wondering if drivers for these lightguns could be reverse-engineered so could be used on all computers with any emulator?

![image](https://github.com/kodi-game/controller-topology-project/assets/6320001/a0643151-fa29-4f14-b016-228ece7bfd7e)

![image](https://github.com/kodi-game/controller-topology-project/assets/6320001/71a69b7c-d64b-4f25-87fd-a70b9edc5747)

![image](https://github.com/kodi-game/controller-topology-project/assets/6320001/71326f20-7790-4644-bd03-813a192c9dc4)

![image](https://github.com/kodi-game/controller-topology-project/assets/6320001/1d316f08-79b0-4389-9bcc-7d4bdc7c1eb6)

![image](https://github.com/kodi-game/controller-topology-project/assets/6320001/c8567549-2de6-41b2-a0dd-1f3529abdb72)

**Comment by Hedda** [comment]
@garbear as I have been on vacation I missed that actually got a reply mail from Andy from Sinden Lightgun a couple of weeks ago saying that he be willing to send you some hardware samples and that he is happy to help you with any requirements for building the software parts if you are interested. You can e-mail him directly at contact@sindenlightgun.com and reference this.

**Comment by Hedda** [comment]
@garbear I now also got a quick reply from **The Retro Shooter Team** who as well wrote that such interest was very well received and that they would very much like to get in contact with you to send you some hardware samples of the new [Retro Shooter lightguns](https://retroshooter.com/). I did not have your e-mail to give them so you must e-mail them directly at retroshooter.info@gmail.com and reference this discussion if you are interested.

**Comment by garbear** [comment]
Sorry, I'm gonna have to decline the offer. My wrist started acting up and I'm trying to limit computer time. Lightgun support would be cool but I just can't put in the computer time like I used to.

**Comment by Hedda** [comment]
Ah, that is sad, hope that you get better soon. 

Maybe some other takers from Team Kodi?
